crypto/aes/aes_ige.c

```diff
@@ -54,6 +54,8 @@ void AES_ige_encrypt(const unsigned char *in, unsigned char *out,
     OPENSSL_assert((length % AES_BLOCK_SIZE) == 0);
```

Comment:
But this assert is going to abort if `length > 0 && length < AES_BLOCK_SIZE`

Reply:
Your right. this assertion in combination with the length == 0 check should be enough, that len (number of AES blocks processed) > 0...

This is the build error i want to address:
```
crypto/aes/aes_ige.c: In function 'AES_ige_encrypt':
crypto/aes/aes_ige.c:135:13: error: 'memcpy' accessing 16 bytes at offsets 0 and 0 overlaps 16 bytes at offset 0 [-Werror=restrict]
  135 |             memcpy(ivec, ivp->data, AES_BLOCK_SIZE);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
crypto/aes/aes_ige.c:82:13: error: 'memcpy' accessing 16 bytes at offsets 0 and 0 overlaps 16 bytes at offset 0 [-Werror=restrict]
   82 |             memcpy(ivec, ivp->data, AES_BLOCK_SIZE);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Well, the code does some pointer aliasing but i can only see this ever happening if the while(len) loop above those lines is skipped (i.e. len==0).
Also adding the proposed changes makes the error disappear, indicating that compiler assumes len==0 can actually happen.

Something is wrong here. Another hint is that i only see this warning in 32 bit build.

Reply:
Looks more like a compiler bug to me.
Is it gcc?

Reply:
```
# gcc --version
gcc (GCC) 9.1.1 20190503 (Red Hat 9.1.1-1)
```
I use the following configuration:
```
/usr/bin/perl ./Configure linux32-s390x --strict-warnings --strict-warnings enable-unit-test enable-md2 enable-rc5  enable-weak-ssl-ciphers enable-ssl3 enable-ssl3-method enable-sctp enable-asan enable-ubsan
```

Reply:
Hmm, using `gcc-Version 10.0.0 20191026 (experimental) (GCC) `
I don't get that far:
```
gcc  -I. -Iinclude -Iproviders/common/include -Iproviders/implementations/include -Icrypto/include  -DAES_ASM -DBSAES_ASM -DECP_NISTZ256_ASM -DGHASH_ASM -DKECCAK1600_ASM -DMD5_ASM -DOPENSSL_BN_ASM_GF2m -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DPOLY1305_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DVPAES_ASM -DWHIRLPOOL_ASM -DX25519_ASM -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wmissing-prototypes -Wstrict-prototypes -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-3\"" -DMODULESDIR="\"/usr/local/lib/ossl-modules\"" -DNDEBUG  -MMD -MF crypto/libcrypto-lib-o_dir.d.tmp -MT crypto/libcrypto-lib-o_dir.o -c -o crypto/libcrypto-lib-o_dir.o crypto/o_dir.c
In file included from /usr/include/string.h:635,
                 from crypto/LPdir_unix.c:42,
                 from crypto/o_dir.c:28:
crypto/LPdir_unix.c: In function 'OPENSSL_DIR_read':
crypto/LPdir_unix.c:134:5: error: '__builtin_strncpy' offset [275, 4095] from the object at 'direntry' is out of the bounds of referenced subobject 'd_name' with type 'char[256]' at offset 19 []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Warray-bounds-Werror=array-bounds]8;;]
  134 |     strncpy((*ctx)->entry_name, direntry->d_name,
      |     ^~~~~~~
In file included from /usr/include/dirent.h:61,
                 from crypto/LPdir_unix.c:44,
                 from crypto/o_dir.c:28:
/usr/include/x86_64-linux-gnu/bits/dirent.h:33:10: note: subobject 'd_name' declared here
   33 |     char d_name[256];  /* We must not include limits.h! */
      |          ^~~~~~
cc1: all warnings being treated as errors
make[1]: *** [crypto/libcrypto-lib-o_dir.o] Error 1
```

Reply:
Looks valid, I need to fix that first...

Reply:
Yeah, using arm-linux-gnueabihf-gcc, gcc-Version 10.0.0 20190907 I get now:
```
arm-linux-gnueabihf-gcc  -Icrypto -Iinclude -Iproviders/common/include  -fPIC -pthread -Wa,--noexecstack -Wall -O3 -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wmissing-prototypes -Wstrict-prototypes -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-3\"" -DMODULESDIR="\"/usr/local/lib/ossl-modules\"" -DNDEBUG  -MMD -MF providers/common/ciphers/libcommon-lib-block.d.tmp -MT providers/common/ciphers/libcommon-lib-block.o -c -o providers/common/ciphers/libcommon-lib-block.o providers/common/ciphers/block.c
arm-linux-gnueabihf-gcc  -Icrypto -Iinclude -Iproviders/common/include  -fPIC -pthread -Wa,--noexecstack -Wall -O3 -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wmissing-prototypes -Wstrict-prototypes -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-3\"" -DMODULESDIR="\"/usr/local/lib/ossl-modules\"" -DNDEBUG  -MMD -MF providers/common/ciphers/libcommon-lib-cipher_ccm.d.tmp -MT providers/common/ciphers/libcommon-lib-cipher_ccm.o -c -o providers/common/ciphers/libcommon-lib-cipher_ccm.o providers/common/ciphers/cipher_ccm.c
In function 'ccm_tls_cipher',
    inlined from 'ccm_cipher_internal' at providers/common/ciphers/cipher_ccm.c:359:16,
    inlined from 'ccm_stream_final' at providers/common/ciphers/cipher_ccm.c:265:9:
providers/common/ciphers/cipher_ccm.c:317:5: error: argument 2 null where non-null expected [-Werror=nonnull]
  317 |     memcpy(ctx->iv + EVP_CCM_TLS_FIXED_IV_LEN, in, EVP_CCM_TLS_EXPLICIT_IV_LEN);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from include/internal/cryptlib.h:14,
                 from providers/common/include/prov/ciphercommon.h:14,
                 from providers/common/ciphers/cipher_ccm.c:12:
providers/common/ciphers/cipher_ccm.c: In function 'ccm_stream_final':
/home/ed/gnu/arm-linux-gnueabihf-linux64/arm-linux-gnueabihf/sys-include/string.h:44:14: note: in a call to function 'memcpy' declared here
   44 | extern void *memcpy (void *__restrict __dest,
      |              ^~~~~~
cc1: all warnings being treated as errors
make[1]: *** [providers/common/ciphers/libcommon-lib-cipher_ccm.o] Error 1
make[1]: Leaving directory `/home/ed/OPC/openssl'
```